### PR TITLE
Added support for dns-search and fixes #30102

### LINKED ISSUE #30102: Cannot connect container to user-defined NAT network on Windows
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**
I am running Docker v1.13.0-rc4 on my Windows 10 client. I specified “bridge”:”none” in the daemon/config.json file and then launched Docker daemon. I then created a user-defined NAT network successfully named mynat.  I successfully launched a nanoserver container which was connected to the mynat network.

I then restarted the docker daemon and noticed that the user-defined nat network is renamed to “nat” – I believe this is by design.

However, I can no longer create a container which is connected to this nat network. Ipconfig is blank.

This has repro’d twice now on my system. Any ideas?

**Steps to reproduce the issue:**
1. Start-Service docker (config/daemon.json has bridge:none)
2. docker network create -d nat --subnet 172.16.0.0/16 --gateway=172.16.0.1 mynat
3. Restart-Service docker
4. docker network ls (Look for nat network - renamed by design as we only support one in WS2016)
5. docker run -it --net=nat nanoserver cmd.exe (and other variants including specifying port mapping, not specifying network name, etc.)

**Describe the results you received:**
No IP Configuration present in the container

**Describe the results you expected:**
IP Configuration should be present in the container

**Additional information you deem important (e.g. issue happens only occasionally):**
Has repro'd multiple times

**Output of `docker version`:**
```
Client:
 Version:      1.13.0-rc4
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Sat Dec 17 01:34:17 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.0-rc4
 API version:  1.25 (minimum version 1.24)
 Go version:   go1.7.3
 Git commit:   88862e7
 Built:        Sat Dec 17 01:34:17 2016
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**
```
Containers: 3
 Running: 0
 Paused: 0
 Stopped: 3
Images: 40
Server Version: 1.13.0-rc4
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 14393 (14393.576.amd64fre.rs1_release_inmarket.161208-2252)
Operating System: Windows 10 Enterprise
OSType: windows
Architecture: x86_64
CPUs: 12
Total Memory: 15.93 GiB
Name: jmesser-dev01
ID: BTUT:DTRE:DVUE:CIX6:CLFW:GJX4:GGV7:GWLQ:WHQQ:OUW5:ZL65:PGMP
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Windows Client (Version 1607 - OS Build 14393.576)

### PR DESCRIPTION
**Pull request by msabansal**
**- What I did**
This PR adds support for --dns-search option for windows. It also fixes https://github.com/docker/docker/issues/30102 Support for dns-search  requires platform support and will only work with the next version of windows.

**- How I did it**

**- How to verify it**
Verified the bug fix using the steps Jason mentioned and also validated that dns-suffix is set on the latest builds.
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by msabansal** [comment]
This will require latest windows platforms to work. It will not throw any error on older builds but the option will be silently ignored. This is the present behavior. Do we want to change it so that specifying --dns-search throws error on unsupported platforms?

File: daemon/container_operations.go [hunk 1]
```diff
@@ -32,12 +32,19 @@ var (
 	getPortMapInfo    = container.GetSandboxPortMapInfo
 )
 
+func (daemon *Daemon) getDNSSearchSettings(container *container.Container) []string {
+	if len(container.HostConfig.DNSSearch) > 0 {
+		return container.HostConfig.DNSSearch
+	} else if len(daemon.configStore.DNSSearch) > 0 {
```
Comment (jhowardmsft) [inline daemon/container_operations.go]: Can remove the else here and just have a second if-statement

File: vendor.conf [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # the following lines are in sorted order, FYI
 github.com/Azure/go-ansiterm 388960b655244e76e24c75f48631564eaefade62
-github.com/Microsoft/hcsshim v0.5.9
+github.com/Microsoft/hcsshim 2c872d137f3a1064686710ec072141b741924fb3
```
Comment (jhowardmsft) [inline vendor.conf]: Can we get this tagged with a version?

File: vendor/github.com/Microsoft/hcsshim/interface.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ type ContainerConfig struct {
 	HvRuntime                *HvRuntime  `json:",omitempty"` // Hyper-V container settings. Used by Hyper-V containers only. Format ImagePath=%root%\BaseLayerID\UtilityVM
 	Servicing                bool        // True if this container is for servicing
 	AllowUnqualifiedDNSQuery bool        // True to allow unqualified DNS name resolution
+	DNSSearchList            string      // Comma seperated list of DNS suffixes to use for name resolution
```
Comment (jhowardmsft) [inline vendor/github.com/Microsoft/hcsshim/interface.go]: Just realized - this should be omitempty. Sorry 😇

Comment (jhowardmsft) [inline vendor/github.com/Microsoft/hcsshim/interface.go]: (Nit, not sure if it makes sense), but consider updating the example at the top of `libcontainerd/client_windows.go` to include this.

File: daemon/daemon_windows.go [hunk 4]
```diff
@@ -340,7 +340,9 @@ func (daemon *Daemon) initNetworkController(config *Config, activeSandboxes map[
 
 		// If there is no nat network create one from the first NAT network
 		// encountered
-		if !defaultNetworkExists && runconfig.DefaultDaemonNetworkMode() == containertypes.NetworkMode(strings.ToLower(v.Type)) {
+		if !defaultNetworkExists &&
+			runconfig.DefaultDaemonNetworkMode() == containertypes.NetworkMode(strings.ToLower(v.Type)) &&
+			n == nil {
```
Comment (jhowardmsft) [inline daemon/daemon_windows.go]: This change looks unrelated. Could you update the comment (or is that still accurate?)

File: libcontainerd/types_windows.go [hunk 5]
```diff
@@ -67,6 +67,11 @@ type CredentialsOption struct {
 	Credentials string
 }
 
+// DNSOption is a CreateOption that provides container specific DNS settings
```
Comment (jhowardmsft) [inline libcontainerd/types_windows.go]: Could/should lthis be combined with the existing `NetworkEndpointsOption` to avoid more proliferations to this pattern of side-band options? Not a problem if it doesn't make sense, just asking the question :innocent:

**Review by lowenna — CHANGES_REQUESTED** [review]

File: daemon/container_operations.go (hunk 1 above)
Comment (msabansal) [inline daemon/container_operations.go]: Accepted

File: vendor.conf (hunk 2 above)
Comment (msabansal) [inline vendor.conf]: will do

File: vendor/github.com/Microsoft/hcsshim/interface.go (hunk 3 above)
Comment (msabansal) [inline vendor/github.com/Microsoft/hcsshim/interface.go]: Updating hcsshim

Comment (msabansal) [inline vendor/github.com/Microsoft/hcsshim/interface.go]: Done

File: daemon/daemon_windows.go (hunk 4 above)
Comment (msabansal) [inline daemon/daemon_windows.go]: The comment includes that it fixes https://github.com/docker/docker/issues/30102. Should I do a separate PR for that?

File: libcontainerd/types_windows.go (hunk 5 above)
Comment (msabansal) [inline libcontainerd/types_windows.go]: Yes it can be. I just thought It would be better to have it as a separate option. I will combine them

File: daemon/daemon_windows.go (hunk 4 above)
Comment (jhowardmsft) [inline daemon/daemon_windows.go]: Ah, I was referring to the comment immediate above this line of code.

File: libcontainerd/types_windows.go (hunk 5 above)
Comment (jhowardmsft) [inline libcontainerd/types_windows.go]: Thanks - yes, I'm trying to figure out how to get rid of side-band options and have everything in OCI.

File: vendor.conf (hunk 2 above)
Comment (jhowardmsft) [inline vendor.conf]: v0.5.10 just released

**Comment by jhowardmsft** [comment]
> This will require latest windows platforms to work. It will not throw any error on older builds but the option will be silently ignored. This is the present behavior. Do we want to change it so that specifying --dns-search throws error on unsupported platforms?

I think that would probably be a good idea.

**Comment by jhowardmsft** [comment]
Does this also require any doc updates?

**Comment by msabansal** [comment]
@jhowardmsft Not sure if it requires a doc update. Perhaps we can add somewhere that dns-search is only supported for rs2. Since we are now throwing an error it should be alright.

**Comment by jhowardmsft** [comment]
LGTM. Thanks for doing the updates @msabansal

**Review by lowenna — APPROVED** [review]

File: daemon/daemon_windows.go [hunk 6]
```diff
@@ -339,8 +339,10 @@ func (daemon *Daemon) initNetworkController(config *Config, activeSandboxes map[
 		name := v.Name
 
 		// If there is no nat network create one from the first NAT network
-		// encountered
-		if !defaultNetworkExists && runconfig.DefaultDaemonNetworkMode() == containertypes.NetworkMode(strings.ToLower(v.Type)) {
+		// encountered if it doesn't already exist
+		if !defaultNetworkExists &&
+			runconfig.DefaultDaemonNetworkMode() == containertypes.NetworkMode(strings.ToLower(v.Type)) &&
+			n == nil {
```
Comment (johnstep) [inline daemon/daemon_windows.go]: Maybe check `n == nil` first, or at least second.

**Review by johnstep — APPROVED** [review]
LGTM

It might be better to separate the fix for #30102, though.

**Comment by msabansal** [comment]
@jhowardmsft @johnstep Can we merge or do I split these?

The test failures don't look related to this change. Can we rerun or should I triage?

File: daemon/daemon_windows.go (hunk 6 above)
Comment (LK4D4) [inline daemon/daemon_windows.go]: ping @msabansal

**Comment by LK4D4** [comment]
@msabansal looks like @johnstep has minor comment for you

**Comment by vdemeester** [comment]
@LK4D4 code looks good and the comment is really a nit. Let's merge this and do a quick follow-up 👼

**Review by vdemeester — APPROVED** [review]
